python/paddle/device/cuda/graphs.py

```diff
@@ -106,3 +109,323 @@ def wrap_cuda_graph(function, mode="thread_local", memory_pool="default"):
+             the logical ins and outs of the cuda graph section
+    """
+    program = paddle.static.Program()
+    block = program.global_block()
+    origin_block = origin_program.global_block()
+    ins = []
+    outs = []
+    op_role_attr_name = core.op_proto_and_checker_maker.kOpRoleAttrName()
+    later_ins = all_inputs_of_later_op(origin_block, section_idx[-1])
+
+    for op in section:
+        for in_name in op.input_arg_names:
+            var = origin_block.var(in_name)
+            new_var_desc = block.desc.var(var.name.encode("ascii"))
+            copy_var_desc(new_var_desc, var)
+            if outs.count(in_name) == 0 and ins.count(in_name) == 0:
+                # This in var is generated from op outside this section
+                # Only record once for same input
+                ins.append(in_name)
+            elif later_ins.count(in_name) == 0:
+                # this is var is generated from op inside this section, and only will be used inside this section
+                outs.remove(in_name)
+        for out_name in op.output_arg_names:
+            var = origin_block.var(out_name)
+            new_var_desc = block.desc.var(var.name.encode("ascii"))
+            copy_var_desc(new_var_desc, var)
+            # for every output, we add it to the section's outs
+            if outs.count(out_name) == 0:
+                # Only record one out var even if it will be generated by multi ops.
+                # For scenario like this:
+                # A = op1(a)
+                # A = op2(b)
+                # B = op3(A)
+                outs.append(out_name)
+        new_op_desc = block.desc.append_op()
+        new_op_desc.copy_from(op.desc)
+        new_op_desc._set_attr(op_role_attr_name, op.attr(op_role_attr_name))
+
+    program._sync_with_cpp()
+
+    return program, [ins, outs]
+
+
+def get_cuda_graph_sections(program):
+    """
+    get all sections that should run under cuda graph and the corresponding idx
+
+    :param program: framework.Program, the original program
+    :return: A list of cuda graph sections and the corresponding ops' idx in the block.
+             The program is under is test or not.
+    """
+    block = program.global_block()
+    cuda_graph_sections = []  # record all ops in every cuda graph sections
+    sections_idx = []  # idx of all ops in every cuda graph sections
+    is_test = False  # will be set to True is any op's 'is_test' attr is True
+
+    # ops and it's idx between cuda graph wrapped op, may belong to a section
+    internal_section = []
+    internal_idx = []
+
+    current_section = []  # current recording cuda graph sections
+    current_idx = []  # current recording cuda graph ops' idx
+    current_cuda_graph_id = -1  # current recording cuda graph id
+    op_role_attr_name = core.op_proto_and_checker_maker.kOpRoleAttrName()
+
+    for idx, op in enumerate(block.ops):
+        assert op.type != 'conditional_block' and op.type != 'while', \
```

Comment:
If the program contains `conditional_block` or `while` ops but these ops are not marked with `wrap_cuda_graph`, should it be OK ?

Reply:
Great point, will be updated to: if the op is `conditional_block` or `while`, it's cuda graph attr must be none.